### PR TITLE
Fix: Truncate UUID to fit varchar(16) column when handling duplicate player names

### PR DESCRIPTION
**Pull request by JobsonMarinho**
## Problem
When a duplicate player name is detected, the code was attempting to set the name to the full UUID string (36 characters), which exceeded the database column limit of varchar(16). This caused a MysqlDataTruncation error.

## Solution
The fix uses a shortened version by taking the first 15 characters of the UUID and prefixing with '~', resulting in a 16-character string that fits within the database constraint.

## Error Fixed
Fixes the error: Data too long for column 'name' at row 1

## Changes
- Truncate UUID to 16 characters (~144d3918-0408-3) instead of using full UUID (144d3918-0408-38f1-b631-8a000d6db3d8) when handling duplicate player names
- File changed: `src/main/java/net/sacredlabyrinth/phaed/simpleclans/listeners/SCPlayerListener.java`

### CODE REVIEW

**Comment by RoinujNosde** [comment]
Increasing the column size is probably a better solution

**Comment by JobsonMarinho** [comment]
> Increasing the column size is probably a better solution

I think the best approach would be to **convert legacy data to support the new account format**.
* If the player **changes their name but keeps the same UUID**, then simply update the `name` column.
* If the **UUID changes but the name stays the same**, then update the `uuid` column instead.

This might not be ideal for every server, but in my case it works well because my network supports **premium, cracked and Bedrock players**. I already handle this logic across all of my plugins.

However, since SimpleClans is a public plugin, I felt it was better to apply a minimal change by adjusting the name to UUID using:
`"~" + duplicate.getUniqueId().toString().substring(0, 15);`

This resolved the warnings/errors that appeared in the console and I thought this solution could be beneficial for all SimpleClans users, not just for my specific setup.

**Comment by RoinujNosde** [comment]
> > Increasing the column size is probably a better solution
> 
> I think the best approach would be to **convert legacy data to support the new account format**.
> 
>     * If the player **changes their name but keeps the same UUID**, then simply update the `name` column.
> 
>     * If the **UUID changes but the name stays the same**, then update the `uuid` column instead.


That would be perfect, indeed.

I tried to implement this once, but had some problems with race conditions IIRC, so I settled for this workaround.

**Comment by JobsonMarinho** [comment]
@RoinujNosde i completely refactored the player name/UUID handling system to properly manage duplicate and outdated player records in the database. This addresses the `MysqlDataTruncation` error that was occurring when duplicate players were detected.

### What Changed
Instead of trying to rename duplicate players with truncated UUIDs (which was causing database errors), we now implement a comprehensive synchronization system that:
1. **Checks for existing records** by both name and UUID  
2. **Handles all edge cases**:
   - Same UUID, different name → Updates the name (player changed their name)  
   - Same name, different UUID → Updates the UUID (player cracked become premium)  
   - Duplicate records with the same name and UUID → Merges them into one  
   - Conflicting records (different name AND different UUID) → Intelligently merges the data  

### Why This Matters
This approach:
- **Prevents data loss**: Instead of deleting or corrupting player data, we merge conflicting records  
- **Fixes the truncation error**: No more trying to fit UUIDs into name columns  
- **Handles edge cases**: Covers scenarios like name changes, server migrations, cracked→premium transitions, or database corruption  
- **Preserves player history**: Keeps the earliest join date and latest activity timestamps when merging  

### Implementation Details
- Added `findClanPlayerByName()` to query players by username  
- Created `syncPlayerData()` to handle all synchronization logic  
- Implemented `mergeClanPlayers()` to intelligently combine duplicate records  
- Updated `updatePlayerName()` to use the new sync system  
- All database operations are properly handled with try-catch blocks  

I think this is the best solution for handling duplicate records. Does this approach align with your vision for SimpleClans? Is there any suggestion or improvement you’d like to see?

**Comment by JobsonMarinho** [comment]
image

**Comment by JobsonMarinho** [comment]
image

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 1]
```diff
@@ -732,19 +616,241 @@ public void run() {
...
+
+    /**
+     * Synchronizes player data in the database, handling duplicates
+     * - If name differs but UUID matches: update name
+     * - If names match but UUIDs differ: update UUID
+     * - If both name and UUID exist in different records: merge and delete duplicates
+     *
+     * @param player the player to sync
+     */
+    public void syncPlayerData(@NotNull Player player) {
```
[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java]: fix: add @NotNull annotations to player parameters for improved null safety

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 2]
```diff
@@ -732,19 +616,241 @@ public void run() {
...
+        }.runTaskAsynchronously(plugin);
+    }
+
+    /**
+     * Retrieves a clan player from the database by name
+     *
+     * @param name the player name to search for
+     * @return the ClanPlayer if found, null otherwise
+     */
+    public @Nullable ClanPlayer retrieveClanPlayerByName(String name) {
```
[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java:730]: fix: add null and empty check for player name retrieval

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 3]
```diff
@@ -732,19 +616,241 @@ public void run() {
...
     }
 
     /**
      * Change the name of a player in the database
      *
      * @param cp to update
+     * @deprecated Use syncPlayerData instead for proper duplicate handling
      */
+    @Deprecated
     public void updatePlayerName(final @NotNull ClanPlayer cp) {
```
[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java:966]: fix: deprecate updatePlayerName method in favor of syncPlayerData for better duplicate handling

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 4]
```diff
@@ -732,19 +616,241 @@ public void run() {
...
+
+        // Merge other properties from primary
+        merged.setFriendlyFire(primary.isFriendlyFire());
+        merged.setFlags(primary.getFlags());
+        merged.setPackedPastClans(primary.getPackedPastClans());
+        merged.setResignTimes(primary.getResignTimes());
+        merged.setLocale(primary.getLocale());
+
+        return merged;
     }
```
[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java:957]: For now, this is not necessary. The logic of mergeClanPlayers method is straightforward enough at this stage. We can consider adding specific unit tests in the future if the merge behavior becomes more complex or if issues arise related to edge cases.

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 5]
```diff
@@ -732,19 +616,241 @@ public void run() {
...
+
+            // Update in-memory
+            plugin.getClanManager().deleteClanPlayerFromMemory(oldByNameUuid);
+            plugin.getClanManager().importClanPlayer(merged);
+            
+            plugin.getLogger().info(String.format("Duplicate records merged for %s (%s)", currentName, currentUuid));
+        } catch (Exception e) {
+            plugin.getServer().getLogger().log(Level.SEVERE, "[SimpleClans] Error synchronizing player data for " + player.getName(), e);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java]: Now, this is not necessary. The current control flow, while complex, reflects the required business logic and is manageable as it stands. Refactoring into smaller helper methods can be considered later if the logic grows further or if maintainability becomes a real concern, but at this point it does not justify the added overhead.

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 6]
```diff
@@ -732,19 +616,241 @@ public void run() {
...
+
+            // Case 3: Found by UUID only, name differs - update name
+            if (byName == null) {
+                plugin.getLogger().info(String.format("Correcting name for %s to %s (%s)", byUuid.getName(), currentName, currentUuid));
+
+                byUuid.setName(currentName);
+                byUuid.setLastSeen(System.currentTimeMillis());
+                updateClanPlayer(byUuid, true); // Force immediate update
+                plugin.getLogger().info(String.format("Player name updated in database: %s", currentName));
+                return;
```
[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java]: fix: enhance player name update logic to prevent UUID truncation issues

[the remark this answers is not in the document]
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java]: fix: enhance player name update logic to prevent UUID truncation issues

**Comment by JobsonMarinho** [comment]
@RoinujNosde can u review it?

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 7]
```diff
@@ -159,74 +111,26 @@ public void initiateDB() {
...
-                    		+ " `fee_enabled` tinyint(1) default '0',"
-                    		+ " `fee_value` double(64,2),"
-                    		+ " `ranks` text NOT NULL,"
-                            + " `banner` text,"
-                    		+ "  PRIMARY KEY  (`id`),"
-                    		+ " UNIQUE (`tag`));";
+                    String query = "CREATE TABLE IF NOT EXISTS `" + getPrefixedTable("clans") + "` (" + " `id` bigint(20)," + " `verified` tinyint(1) default '0'," + " `tag` varchar(25) NOT NULL," + " `color_tag` varchar(255) NOT NULL," + " `name` varchar(100) NOT NULL," + " `description` varchar(255)," + " `friendly_fire` tinyint(1) default '0'," + " `founded` bigint NOT NULL," + " `last_used` bigint NOT NULL," + " `packed_allies` text NOT NULL," + " `packed_rivals` text NOT NULL," + " `packed_bb` mediumtext NOT NULL," + " `cape_url` varchar(255) NOT NULL," + " `flags` text NOT NULL," + " `balance` double(64,2) default 0.0," + " `fee_enabled` tinyint(1) default '0'," + " `fee_value` double(64,2)," + " `ranks` text NOT NULL," + " `banner` text," + "  PRIMARY KEY  (`id`)," + " UNIQUE (`tag`));";
                     core.execute(query);
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java:190]: can you reverse the line breaks? it's better for reading IMO

**Review by RoinujNosde — CHANGES_REQUESTED** [review]

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java (hunk 7 above)
Comment (JobsonMarinho) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java:190]: @RoinujNosde claro, me desculpa meu monitor é ultra wide, irei corrigir

**Comment by JobsonMarinho** [comment]
@RoinujNosde I reversed the line breaks, anything else?

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 8]
```diff
@@ -732,19 +712,245 @@ public void run() {
...
+     *
+     * @param player the player to sync
+     */
+    public void syncPlayerData(@NotNull Player player) {
+        try {
+            String currentName = player.getName();
+            UUID currentUuid = player.getUniqueId();
+
+            ClanPlayer byName = retrieveClanPlayerByName(currentName);
+            ClanPlayer byUuid = retrieveOneClanPlayer(currentUuid);
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java:834]: This is calling database code in the main thread, it's going to cause lag.

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java [hunk 9]
```diff
@@ -732,19 +712,245 @@ public void run() {
...
+
+            // Case 1: No records exist - create new
+            if (byName == null && byUuid == null) {
+                plugin.getLogger().info(String.format("No existing records for %s (%s)", currentName, currentUuid));
+                return;
+            }
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/managers/StorageManager.java]: This will never happen. The object is already created here: https://github.com/RoinujNosde/SimpleClans/blob/90b7378ee3525236ce273f9b7756d86561a5509b/src/main/java/net/sacredlabyrinth/phaed/simpleclans/listeners/SCPlayerListener.java#L86
